### PR TITLE
Refactor: Remove old `abs` runtime implementations

### PR DESCRIPTION
**Pull request by Thirumalai-Shaktivel**
Towards: https://github.com/lfortran/lfortran/issues/3034

### CODE REVIEW

**Comment by Thirumalai-Shaktivel** [comment]
TODO:
- [ ] Map iabs
- [ ] Fix 
```
/usr/bin/ld: ../src/libminpack.a(hybrd.f.o): in function `_lcompilers_abs_f64':
LFortran:(.text+0x0): multiple definition of `_lcompilers_abs_f64'; CMakeFiles/example_hybrd.dir/example_hybrd.f90.o:LFortran:(.text+0x0): first defined here
````

**Comment by certik** [comment]
Awesome, thanks for working on this.

**Comment by Thirumalai-Shaktivel** [comment]
```
/usr/bin/ld: ../src/libminpack.a(hybrd.f.o): in function `_lcompilers_abs_f64':
LFortran:(.text+0x0): multiple definition of `_lcompilers_abs_f64'; CMakeFiles/example_hybrd.dir/example_hybrd.f90.o:LFortran:(.text+0x0): first defined here
```
For the above error, I think the issue is that the intrinsic functions that we create are not enclosed within a module (i.e., the file contains subroutines) and get clashed with the other files.
@certik I have seen this before, but I'm not able to trace it. Is there a fix for this?

I think one way to fix this is to create a module called `intrinsic_scalar_functions` and declare all the functions there, and then import wherever required. Do you see any downside to this?

**Comment by certik** [comment]
I know we discussed this issue. One option is to put them into a module and keep importing from it. Another is to keep them global.

The issue with SciPy is that we compile each file into an object file. In other compilers these functions are typically part of their runtime library that they link at the end.

If we create a module, we would have a `.mod` file for each object file, and each separate file might generate slightly different `.mod` file, then we need to merge them etc.

I think the solution we decided to go with is to use name mangling. We have implemented quite sophisticated name mangling mechanisms. I think we keep these functions global, but we mangle global non `bind(c)` functions, that way they will not clash.

**Comment by certik** [comment]
The best is to try to reproduce this error locally, and then figure out the right command line option to fix it.

File: src/libasr/pass/unique_symbols.cpp [hunk 1]
```diff
@@ -79,6 +79,12 @@ class SymbolRenameVisitor: public ASR::BaseWalkVisitor<SymbolRenameVisitor> {
         current_scope = xx.m_symtab;
         std::unordered_map<ASR::symbol_t*, std::string> tmp_scope;
         for (auto &a : xx.m_symtab->get_scope()) {
+            if (global_symbols_mangling && ASR::is_a<ASR::Function_t>(*a.second)) {
+                char *fn_name = ASR::down_cast<ASR::Function_t>(a.second)->m_name;
+                if (startswith(fn_name, "_lcompilers_")) {
+                    should_mangle = true;
+                }
+            }
```
Comment (certik) [inline src/libasr/pass/unique_symbols.cpp:87]: This can be removed if the CI passes now.

**Comment by certik** [comment]
So it now fails with:
```
LFortran:(.text+0x715): undefined reference to `__module_lfortran_intrinsic_math2_lfortran_intrinsic_math2_dmax1_7hrPjB8NSIaUCGqOMr4G58'
/usr/bin/ld: LFortran:(.text+0x943): undefined reference to `__module_lfortran_intrinsic_math2_lfortran_intrinsic_math2_dmin1_7hrPjB8NSIaUCGqOMr4G58'
/usr/bin/ld: LFortran:(.text+0xbb6): undefined reference to `__module_lfortran_intrinsic_math2_lfortran_intrinsic_math2_dmax1_7hrPjB8NSIaUCGqOMr4G58'
/usr/bin/ld: LFortran:(.text+0xdd7): undefined reference to `__module_lfortran_intrinsic_math2_lfortran_intrinsic_math2_dmax1_7hrPjB8NSIaUCGqOMr4G58'
/usr/bin/ld: ../src/libminpack.a(qrfac.f.o): in function `qrfac':
LFortran:(.text+0x598): undefined reference to `__module_lfortran_intrinsic_math2_lfortran_intrinsic_math2_dmax1_YJwo04SE6H0D4LpSYqhTUu'
/usr/bin/ld: ../src/libminpack.a(fdjac1.f.o): in function `fdjac1':
LFortran:(.text+0xa3): undefined reference to `__module_lfortran_intrinsic_math2_lfortran_intrinsic_math2_dmax1_spDbfTsccL6tR5yxHOJO0B'
/usr/bin/ld: ../src/libminpack.a(dogleg.f.o): in function `dogleg':
LFortran:(.text+0x212): undefined reference to `__module_lfortran_intrinsic_math2_lfortran_intrinsic_math2_dmax1_2A6nnSzZOZSEh0RVCdVAGy'
```
I think this is due to our enabling of `--intrinsic-mangling`, which mangles our own old "intrinsics" library (which we are in the process of removing).

I think we need to keep the old library, only mangle `_lcompilers_`.

**Comment by Thirumalai-Shaktivel** [comment]
Cool, yup, that works for Legacy Minpack.

**Comment by certik** [comment]
Yes, legacy minpack now works. 

This is relevant to fix later: https://github.com/lfortran/lfortran/issues/3268.

**Comment by Thirumalai-Shaktivel** [comment]
Ready!

**Review by certik — APPROVED** [review]
I think this looks good now.

**Pull request merged**